README.md
Usermanager
===================

Please, help translating: https://www.transifex.com/solydxk/usermanager/dashboard/

Manage users on your system.


debian/changelog
usermanager (0.8.1) debian; urgency=low

  * Get config from /etc/adduser.conf
  * Set shell for users
  * Do not show face for system users

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sat, 09 Apr 2016 15:11:34 +0200

usermanager (0.8.0) debian; urgency=low

  * Add show system users

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sat, 09 Apr 2016 13:38:19 +0200

usermanager (0.7.9) debian; urgency=low

  * Updated translations

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Mon, 18 Jan 2016 18:00:00 +0100

usermanager (0.7.8) debian; urgency=low

  * Updated translations

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sat, 19 Dec 2015 11:00:00 +0100

usermanager (0.7.6) debian; urgency=low

  * Updated translations

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Thu, 26 Nov 2015 12:00:00 +0100

usermanager (0.7.4) debian; urgency=low

  * Added translations: de, it

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Mon, 19 Oct 2015 07:00:00 +0100

usermanager (0.7.3) debian; urgency=low

  * Improved localization handling

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Mon, 22 June 2015 12:00:00 +0100

usermanager (0.7.2) debian; urgency=low

  * Added KCM launcher for System Settings

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Fri, 08 May 2015 17:00:00 +0100

usermanager (0.7.1) debian; urgency=low

  * Added X-XFCE-SettingsDialog;X-XFCE-SystemSettings; to launcher

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Wed, 06 May 2015 15:00:00 +0100

usermanager (0.7.0) debian; urgency=low

  * Lowered gtk+ version for Wheezy

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sat, 21 Jan 2015 13:00:00 +0100

usermanager (0.6.9) debian; urgency=low

  * Minor GUI improvements

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Thu, 01 Jan 2015 17:00:00 +0100

usermanager (0.6.8) debian; urgency=low

  * Improved translation handling

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sun, 02 Nov 2014 10:00:00 +0100

usermanager (0.6.7) debian; urgency=low

  * Changed icon for Evolvere icon theme

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Wed, 17 Sep 2014 19:00:00 +0100

usermanager (0.6.6) debian; urgency=low

  * Moved locale to solydxk-locale

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Thu, 26 May 2014 08:00:00 +0100

usermanager (0.6.5) debian; urgency=low

  * Bug resolved with /dev/null as home directory

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Wed, 30 Apr 2014 11:00:00 +0100

usermanager (0.6.4) debian; urgency=low

  * Added German, Greek, Dutch, and Turkish translations

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Wed, 02 Apr 2014 19:00:00 +0100

usermanager (0.6.3) debian; urgency=low

  * Resolved bug not being able to save real user names with spaces

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Sat, 30 Nov 2013 18:00:00 +0100

usermanager (0.5.9) debian; urgency=low

  * Initial version

 -- Arjen Balfoort <arjenbalfoort@solydxk.com>  Thu, 03 Oct 2013 13:00:00 +0100


usr/lib/usermanager/dialogs.py
#! /usr/bin/env python3

# Make sure the right Gtk version is loaded
import gi
gi.require_version('Gtk', '3.0')

from gi.repository import Gtk, GObject, GdkPixbuf

# Show message dialog
# Usage:
# MessageDialog(_("My Title"), "Your (error) message here", Gtk.MessageType.ERROR).show()
# Message types:
# Gtk.MessageType.INFO
# Gtk.MessageType.WARNING
# Gtk.MessageType.ERROR
# MessageDialog can be called from a working thread
class MessageDialog(Gtk.MessageDialog):
    def __init__(self, title, message, style, parent=None):
        Gtk.MessageDialog.__init__(self, parent, Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT, style, Gtk.ButtonsType.OK, message)
        self.set_default_response(Gtk.ResponseType.OK)
        self.set_position(Gtk.WindowPosition.CENTER)
        self.set_markup("<b>%s</b>" % title)
        self.format_secondary_markup(message)
        if parent is not None:
            self.set_icon(parent.get_icon())
        self.connect('response', self._handle_clicked)

    def _handle_clicked(self, *args):
        self.destroy()

    def show(self):
        GObject.timeout_add(0, self._do_show_dialog)

    def _do_show_dialog(self):
        self.show_all()
        return False


# Show unthreaded message dialog
# Usage:
# MessageDialog(_("My Title"), "Your (error) message here", Gtk.MessageType.ERROR).show()
# Message types:
# Gtk.MessageType.INFO
# Gtk.MessageType.WARNING
# Gtk.MessageType.ERROR
# MessageDialogSave can NOT be called from a working thread, only from main (UI) thread
class MessageDialogSave(object):
    def __init__(self, title, message, style, parent=None):
        self.title = title
        self.message = message
        self.parent = parent
        self.style = style

    def show(self):
        dialog = Gtk.MessageDialog(self.parent, Gtk.DialogFlags.MODAL, self.style, Gtk.ButtonsType.OK, self.message)
        dialog.set_markup("<b>%s</b>" % self.title)
        dialog.format_secondary_markup(self.message)
        if self.parent is not None:
            dialog.set_icon(self.parent.get_icon())
        dialog.run()
        dialog.destroy()


# Create question dialog
# Usage:
# dialog = QuestionDialog(_("My Title"), _("Put your question here?"))
#    if (dialog.show()):
# QuestionDialog can NOT be called from a working thread, only from main (UI) thread
class QuestionDialog(object):
    def __init__(self, title, message, parent=None):
        self.title = title
        self.message = message
        self.parent = parent

    #''' Show me on screen '''
    def show(self):
        dialog = Gtk.MessageDialog(self.parent, Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT, Gtk.MessageType.QUESTION, Gtk.ButtonsType.YES_NO, self.message)
        dialog.set_markup("<b>%s</b>" % self.title)
        dialog.format_secondary_markup(self.message)
        dialog.set_position(Gtk.WindowPosition.CENTER)
        if self.parent is not None:
            dialog.set_icon(self.parent.get_icon())
        answer = dialog.run()
        if answer == Gtk.ResponseType.YES:
            return_value = True
        else:
            return_value = False
        dialog.destroy()
        return return_value


class SelectImageDialog(object):
    def __init__(self, title, start_directory=None, parent=None):
        self.title = title
        self.start_directory = start_directory
        self.parent = parent

    def show(self):
        imagePath = None
        image = Gtk.Image()

        # Image preview function
        def image_preview_cb(dialog):
            filename = dialog.get_preview_filename()
            try:
                pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(filename, 128, 128)
                image.set_from_pixbuf(pixbuf)
                valid_preview = True
            except:
                valid_preview = False
            dialog.set_preview_widget_active(valid_preview)

        dialog = Gtk.FileChooserDialog(self.title, self.parent, Gtk.FileChooserAction.OPEN, (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OPEN, Gtk.ResponseType.OK))
        dialog.set_default_response(Gtk.ResponseType.OK)
        if self.start_directory is not None:
            dialog.set_current_folder(self.start_directory)
        fleFilter = Gtk.FileFilter()
        fleFilter.set_name("Images")
        fleFilter.add_mime_type("image/png")
        fleFilter.add_mime_type("image/jpeg")
        fleFilter.add_mime_type("image/gif")
        fleFilter.add_pattern("*.png")
        fleFilter.add_pattern("*.jpg")
        fleFilter.add_pattern("*.gif")
        fleFilter.add_pattern("*.tif")
        fleFilter.add_pattern("*.xpm")
        dialog.add_filter(fleFilter)

        # Add a preview widget:
        dialog.set_preview_widget(image)
        dialog.connect("update-preview", image_preview_cb)

        answer = dialog.run()
        if answer == Gtk.ResponseType.OK:
            imagePath = dialog.get_filename()
        dialog.destroy()
        return imagePath


class SelectDirectoryDialog(object):
    def __init__(self, title, start_directory=None, parent=None):
        self.title = title
        self.start_directory = start_directory
        self.parent = parent

    def show(self):
        directory = None
        dialog = Gtk.FileChooserDialog(self.title, self.parent, Gtk.FileChooserAction.SELECT_FOLDER, (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OPEN, Gtk.ResponseType.OK))
        dialog.set_default_response(Gtk.ResponseType.OK)
        if self.start_directory is not None:
            dialog.set_current_folder(self.start_directory)
        answer = dialog.run()
        if answer == Gtk.ResponseType.OK:
            directory = dialog.get_filename()
        dialog.destroy()
        return directory


usr/lib/usermanager/execcmd.py
#! /usr/bin/env python3
#-*- coding: utf-8 -*-

import sys
import subprocess
#import gettext

# i18n: http://docs.python.org/2/library/gettext.html
#t = gettext.translation("solydxk-conky", "/usr/share/locale")
#_ = t.lgettext


# Class to execute a command and return the output in an array
class ExecCmd(object):

    def __init__(self, loggerObject=None):
        self.log = loggerObject

    def run(self, cmd, realTime=True, returnAsList=True):
        if self.log:
            self.log.write(_("Command to execute: %(cmd)s") % { "cmd": cmd }, 'execcmd.run', 'debug')

        p = subprocess.Popen([cmd], shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        lstOut = []
        while True:
            line = p.stdout.readline()
            if not line:
                break
            # Strip the line, also from null spaces (strip() only strips white spaces)
            line = line.decode('utf-8').strip().strip("\0")
            lstOut.append(line)
            if realTime:
                sys.stdout.flush()
                if self.log:
                    self.log.write(line, 'execcmd.run', 'info')

        ret = lstOut
        if not returnAsList:
            ret = "\n".join(lstOut)
        return ret


usr/lib/usermanager/combobox.py
#! /usr/bin/env python3

# Make sure the right Gtk version is loaded
import gi
gi.require_version('Gtk', '3.0')

from gi.repository import Gtk, GObject


class ComboBoxHandler(GObject.GObject):

    def __init__(self, combobox, loggerObject=None):
        GObject.GObject.__init__(self)
        self.combobox = combobox
        self.log = loggerObject

    # Clear treeview
    def clearComboBox(self):
        liststore = self.combobox.get_model()
        if liststore is not None:
            liststore.clear()
            self.combobox.set_model(liststore)
        self.setValue("")

    def fillComboBox(self, dataList, select_value=None):
        if dataList:
            self.clearComboBox()
            liststore = self.combobox.get_model()
            if liststore is None:
                liststore = Gtk.ListStore(str)
                if self.combobox.get_has_entry():
                    entry = self.combobox.get_child()
                    entry.set_can_focus(True)
                else:
                    cell = Gtk.CellRendererText()
                    self.combobox.pack_start(cell, True)
                    self.combobox.add_attribute(cell, "text", 0)
            for data in dataList:
                liststore.append([str(data)])
            self.combobox.set_model(liststore)
            if select_value is not None:
                self.selectValue(select_value)

    def selectValue(self, value, valueColNr=0):
        i = 0
        activeIndex = -1
        liststore = self.combobox.get_model()
        if liststore:
            for data in liststore:
                if data[valueColNr] == value:
                    activeIndex = i
                    break
                i += 1
            if self.combobox.get_has_entry():
                self.combobox.set_entry_text_column(valueColNr)
                if activeIndex < 0:
                    self.setValue(value)
            else:
                if activeIndex < 0:
                    activeIndex = 0
            self.combobox.set_active(activeIndex)

    def setValue(self, value):
        if self.combobox.get_has_entry():
            entry = self.combobox.get_child()
            entry.set_text(value)

    def getValue(self):
        value = None
        if self.combobox.get_has_entry():
            entry = self.combobox.get_child()
            value = entry.get_text().strip()
        else:
            model = self.combobox.get_model()
            value = model.get_value(self.combobox.get_active_iter(), 0)
        return value

# Register the class
GObject.type_register(ComboBoxHandler)


usr/lib/usermanager/user.py
#! /usr/bin/env python3

# Make sure the right Gtk version is loaded
import gi
gi.require_version('Gtk', '3.0')

import os
import pwd
import spwd
import grp
import operator
import crypt
from shutil import copy
from datetime import datetime, timedelta
from gi.repository import GdkPixbuf, Gtk
from os.path import join, exists
from functions import get_config_dict

# i18n: http://docs.python.org/3/library/gettext.html
import gettext
from gettext import gettext as _
gettext.textdomain('usermanager')

#pwd
#Index	Attribute	Meaning
#0	pw_name	Login name
#1	pw_passwd	Optional encrypted password
#2	pw_uid	Numerical user ID
#3	pw_gid	Numerical group ID
#4	pw_gecos	User name or comment field
#5	pw_dir	User home directory
#6	pw_shell	User command interpreter

#pwd.getpwuid(uid)
#Return the password database entry for the given numeric user ID.

#pwd.getpwnam(name)
#Return the password database entry for the given user name.

#pwd.getpwall()
#Return a list of all available password database entries, in arbitrary order.

#==============================================================================

#spwd
#Index	Attribute	Meaning
#0	sp_nam	Login name
#1	sp_pwd	Encrypted password
#2	sp_lstchg	Date of last change
#3	sp_min	Minimal number of days between changes
#4	sp_max	Maximum number of days between changes
#5	sp_warn	Number of days before password expires to warn user about it
#6	sp_inact	Number of days after password expires until account is blocked
#7	sp_expire	Number of days since 1970-01-01 until account is disabled
#8	sp_flag	Reserved

#spwd.getspnam(name)
#Return the shadow password database entry for the given user name.

#spwd.getspall()
#Return a list of all available shadow password database entries, in arbitrary order.

#==============================================================================

#grp
#Index	Attribute	Meaning
#0	gr_name	the name of the group
#1	gr_passwd	the (encrypted) group password; often empty
#2	gr_gid	the numerical group ID
#3	gr_mem	all the group member’s user names

#grp.getgrgid(gid)
#Return the group database entry for the given numeric group ID. KeyError is raised if the entry asked for cannot be found.

#grp.getgrnam(name)
#Return the group database entry for the given group name. KeyError is raised if the entry asked for cannot be found.

#grp.getgrall()
#Return a list of all available group entries, in arbitrary order.


class User(object):

    def __init__(self, loggerObject=None):
        self.log = loggerObject
        conf = "/etc/adduser.conf"
        # Set conf variables
        config = get_config_dict(conf)
        self.first_uid = int(config.get('FIRST_UID', 1000))
        self.last_uid = int(config.get('LAST_UID', 29999))
        self.first_gid = int(config.get('FIRST_GID', 1000))
        self.last_gid = int(config.get('LAST_GID', 29999))

    def getAllUsersInfoDict(self, include_system_users=False):
        users = []
        pwds = self.sortListOnColumn(pwd.getpwall(), [0])
        for p in pwds:
            # Check UID range when returning non-system users only
            addUser = True
            if not include_system_users:
                if p.pw_uid < self.first_uid or p.pw_uid > self.last_uid:
                    addUser = False
            if addUser:
                users.append({'user': p,
                              'groups': self.getUserGroups(p.pw_name),
                              'prgrp': self.getUserPrimaryGroupName(p.pw_name),
                              'pwd': self.getUserPasswordInfoDict(p.pw_name),
                              'face': self.getUserFacePath(p.pw_name)})
        return users

    def getUserGroups(self, name):
        groups = [g.gr_name for g in grp.getgrall() if name in g.gr_mem]
        gid = pwd.getpwnam(name).pw_gid
        groups.append(grp.getgrgid(gid).gr_name)
        return groups

    def getGroupAccounts(self, group):
        groupAccounts = []
        try:
            g = grp.getgrnam(group)
            gid = g.gr_gid
            pwds = pwd.getpwall()
            for p in pwds:
                if p.pw_gid == gid:
                    groupAccounts.append(p.pw_name)
            for u in g.gr_mem:
                groupAccounts.append(u)
        except:
            # Best effort
            pass
        return groupAccounts

    def getGroups(self):
        groups = []
        groupsDict = grp.getgrall()
        for group in groupsDict:
            groups.append(group.gr_name)
        groups.sort()
        return groups

    def doesGroupExist(self, group):
        return group in self.getGroups()

    def getUsers(self, include_system_users=False):
        users = []
        pwds = pwd.getpwall()
        for p in pwds:
            addUser = True
            if not include_system_users:
                if p.pw_uid < self.first_uid or p.pw_uid > self.last_uid:
                    addUser = False
            if addUser:
                users.append(p.pw_name)
        users.sort()
        return users

    def getLoggedinUser(self):
        p = os.popen('logname', 'r')
        userName = p.readline().strip()
        p.close()
        if userName is "":
            userName = pwd.getpwuid(os.getuid()).pw_name
        return userName

    def doesUserExist(self, user):
        return user in self.getUsers(True)

    def getUserPrimaryGroupName(self, name):
        p = pwd.getpwnam(name)
        return grp.getgrgid(p.pw_gid).gr_name

    def getUserHomeDir(self, name):
        return pwd.getpwnam(name).pw_dir

    def getUserFacePath(self, name):
        face = None
        if not self.isSystemUser(name):
            homeDir = self.getUserHomeDir(name)
            if exists(homeDir):
                # Check for face icon
                if exists(join(homeDir, ".face")):
                    face = join(homeDir, ".face")
                elif exists(join(homeDir, ".face.icon")):
                    face = join(homeDir, ".face.icon")

                if face is None:
                    genericFace = '/usr/share/pixmaps/faces/user-generic.png'
                    kdeFace = '/usr/share/kde4/apps/kdm/faces/.default.face.icon'
                    if exists(genericFace):
                        face = genericFace
                    elif exists(kdeFace):
                        face = kdeFace
                    else:
                        defaultTheme = Gtk.IconTheme.get_default()
                        iconInfo = Gtk.IconTheme.lookup_icon(defaultTheme, "user-identity", 64, Gtk.IconLookupFlags.NO_SVG)
                        if iconInfo is not None:
                            face = iconInfo.get_filename()
                    if face is not None:
                        try:
                            copy(face, join(homeDir, ".face"))
                        except:
                            # Best effort
                            face = None
        return face

    def isSystemUser(self, name):
        pwds = pwd.getpwall()
        for p in pwds:
            if p.pw_name == name:
                if p.pw_uid < self.first_uid or p.pw_uid > self.last_uid:
                    return True
        return False

    def getUserFacePixbuf(self, name=None, width=None, height=None):
        pb = None
        if name is None:
            name = pwd.getpwuid(os.getuid()).pw_name
        if not self.isSystemUser(name):
            facePath = self.getUserFacePath(name)
            if facePath is None:
                defaultTheme = Gtk.IconTheme.get_default()
                iconInfo = Gtk.IconTheme.lookup_icon(defaultTheme, "user-identity", 64, Gtk.IconLookupFlags.NO_SVG)
                if iconInfo is not None:
                    facePath = iconInfo.get_filename()
            if facePath is not None:
                pb = GdkPixbuf.Pixbuf.new_from_file(facePath)
                if width is not None and height is not None:
                    pb.scale_simple(width, height, GdkPixbuf.InterpType.BILINEAR)
                elif width is not None:
                    height = pb.get_height() * (width / pb.get_width())
                    pb.scale_simple(width, height, GdkPixbuf.InterpType.BILINEAR)
                elif height is not None:
                    width = pb.get_width() * (height / pb.get_height())
                    pb.scale_simple(width, height, GdkPixbuf.InterpType.BILINEAR)
            return pb

    def getUserPasswordInfoDict(self, name):
        return spwd.getspnam(name)

    def getUserID(self, name):
        return pwd.getpwdwnam(name).pw_uid

    def getNewUserID(self):
        pwds = pwd.getpwall()
        for new_id in range(self.first_uid, self.last_uid):
            for p in pwds:
                if p.pw_uid != new_id:
                    return new_id
        return 0

    def getGroupID(self, group):
        return grp.getgrnam(group).gr_gid

    def getNewGroupID(self):
        grps = grp.getgrall()
        for new_id in range(self.first_gid, self.last_gid):
            for g in grps:
                if g.gr_gid != new_id:
                    return new_id
        return 0

    def getShells(self):
        shells_path = '/etc/shells'
        shells = []
        if exists(shells_path):
            with open(shells_path, 'r') as f:
                lines = f.readlines()
            for line in lines:
                line = line.strip()
                if exists(line):
                    shells.append(line)
        if not shells:
            for shell in ['/bin/bash', '/bin/sh', '/bin/ksh', '/bin/dash',
                          '/bin/rbash', '/bin/ksh93', '/usr/bin/ksh']:
                if exists(shell):
                    shells.append(shell)
        return shells

    def createGroup(self, group):
        cmd = "addgroup %s" % group
        print((">>> createGroup = %s" % cmd))
        return os.system(cmd)

    def deleteGroup(self, group):
        retMsg = ""
        cmd = "delgroup %s" % group
        print((">>> delgroup = %s" % cmd))
        ret = os.system(cmd)
        if ret > 0:
            if ret == 1280 or ret == 1792:
                retMsg = _("Cannot remove group '%(group)s': not empty." % {"group": group})
            else:
                retMsg = _("The command '%(cmd)s' returned error code: %(ret)d." % {"cmd": cmd, "ret": ret})
        return retMsg

    def removeGroupFromAccount(self, user, group):
        newGroups = []
        userGroups = self.getUserGroups(user)
        for ug in userGroups:
            if ug != group:
                newGroups.append(ug)
        if newGroups:
            self.manageUser(user, "", newGroups)

    def addGroupToAccount(self, user, group):
        cmd = "usermod -aG %(group)s %(user)s" % {"group": group, "user": user}
        print((">>> addGroupToAccount = %s" % cmd))
        return os.system(cmd)

    def manageUser(self, user, primary_group="", group_list=[], shell="", home_dir="", full_name="", password="", expire_date="", inactive_days=""):
        groups = ""

        user_cmd = "useradd"
        if self.doesUserExist(user):
            user_cmd = "usermod"

        if primary_group != "":
            if not self.doesGroupExist(primary_group):
                self.createGroup(primary_group)
            primary_group = "-g %(primary_group)s" % {"primary_group": primary_group}

        if group_list:
            for grp in group_list:
                if not self.doesGroupExist(grp):
                    self.createGroup(grp)
            groups = "-G %s" % ",".join(group_list)

        if shell != "":
            shell = "-s %(shell)s" % {"shell": shell}

        enc_password = ""
        if password != "":
            enc_password = self.encryptPassword(password)
            enc_password = "-p %(enc_password)s" % {"enc_password": enc_password}

        if home_dir != "":
            home_dir = "-d '%(home_dir)s'" % {"home_dir": home_dir}
            user = "-m %(user)s" % {"user": user}

        if full_name != "":
            full_name = "-c '%(full_name)s'" % {"full_name": full_name}

        if expire_date != "":
            expire_date = "-e %(expire_date)s" % {"expire_date": expire_date}

        if inactive_days != "":
            inactive_days = "-f %(inactive_days)s" % {"inactive_days": inactive_days}

        cmd = "%(user_cmd)s %(primary_group)s %(groups)s %(shell)s %(enc_password)s %(home_dir)s \
        %(full_name)s %(expire_date)s %(inactive_days)s %(user)s" % \
        {"user_cmd": user_cmd, "primary_group": primary_group, "groups": groups, "shell": shell, "enc_password": enc_password, \
        "home_dir": home_dir, "full_name": full_name, "expire_date": expire_date, "inactive_days": inactive_days, "user": user}
        print((">>> manageUser = %s" % cmd))
        return os.system(cmd)

    def deleteUser(self, user):
        retMsg = ""
        cmd = "deluser --remove-home %s" % user
        print((">>> deleteUser = %s" % cmd))
        ret = os.system(cmd)
        if ret > 0:
            if ret == 256:
                retMsg = _("Cannot remove user '%(user)s': processes started by user are still running." % {"user": user})
            else:
                retMsg = _("The command '%(cmd)s' returned error code: %(ret)d." % {"cmd": cmd, "ret": ret})
        return retMsg

    # Convert integer to (formatted) date
    def intToDate(self, nr_days, format_string=None, start_date=None):
        if start_date is None:
            start_date = datetime.now()
        dt = start_date + timedelta(days=nr_days)
        if format_string is None:
            return dt
        else:
            return dt.strftime(format_string)

    def disableUserAccount(self, user):
        cmd = "usermod -L -e 1970-01-01 %s" % user
        print((">>> disableUserAccount = %s" % cmd))
        return os.system(cmd)

    def enableUserAccount(self, user):
        cmd = "usermod -U -e 1970-01-01 %s" % user
        print((">>> enableUserAccount = %s" % cmd))
        return os.system(cmd)

    def encryptPassword(self, password):
        return crypt.crypt(password, "22")

    # Sort list on given column
    def sortListOnColumn(self, lst, columsList):
        for col in reversed(columsList):
            lst = sorted(lst, key=operator.itemgetter(col))
        return lst
